src/lib/statistics/chunk_statistics/histograms/equal_height_histogram.cpp

```diff
@@ -0,0 +1,199 @@
+    return std::make_shared<EqualHeightHistogram<T>>(bin_stats.maxs, bin_stats.distinct_counts, bin_stats.min,
+                                                     bin_stats.total_count, characters, prefix_length);
+  } else {
+    DebugAssert(!static_cast<bool>(supported_characters) && !static_cast<bool>(string_prefix_length),
+                "Do not provide string prefix prefix arguments for non-string histograms.");
+    return std::make_shared<EqualHeightHistogram<T>>(bin_stats.maxs, bin_stats.distinct_counts, bin_stats.min,
+                                                     bin_stats.total_count);
+  }
+}
+
+template <typename T>
+HistogramType EqualHeightHistogram<T>::histogram_type() const {
+  return HistogramType::EqualHeight;
+}
+
+template <typename T>
+BinID EqualHeightHistogram<T>::bin_count() const {
+  return _maxs.size();
+}
+
+template <typename T>
+BinID EqualHeightHistogram<T>::_bin_for_value(const T value) const {
+  if (value < _min) {
+    return INVALID_BIN_ID;
+  }
+
+  const auto it = std::lower_bound(_maxs.cbegin(), _maxs.cend(), value);
+
+  if (it == _maxs.cend()) {
+    return INVALID_BIN_ID;
+  }
+
+  return static_cast<BinID>(std::distance(_maxs.cbegin(), it));
+}
+
+template <typename T>
+BinID EqualHeightHistogram<T>::_upper_bound_for_value(const T value) const {
+  const auto it = std::upper_bound(_maxs.cbegin(), _maxs.cend(), value);
+
+  if (it == _maxs.cend()) {
+    return INVALID_BIN_ID;
+  }
+
+  return static_cast<BinID>(std::distance(_maxs.cbegin(), it));
+}
+
+template <typename T>
+T EqualHeightHistogram<T>::_bin_min(const BinID index) const {
+  DebugAssert(index < this->bin_count(), "Index is not a valid bin.");
+
+  // If it's the first bin, return _min.
+  if (index == 0u) {
+    return _min;
+  }
+
+  // Otherwise, return the next representable value of the previous bin's max.
+  return this->_get_next_value(this->_bin_max(index - 1));
+}
+
+template <typename T>
+T EqualHeightHistogram<T>::_bin_max(const BinID index) const {
+  DebugAssert(index < _maxs.size(), "Index is not a valid bin.");
+  return _maxs[index];
+}
+
+template <typename T>
+HistogramCountType EqualHeightHistogram<T>::_bin_height(const BinID index) const {
+  DebugAssert(index < this->bin_count(), "Index is not a valid bin.");
+  return total_count() / bin_count() + (total_count() % bin_count() > 0 ? 1 : 0);
+}
+
+template <typename T>
+HistogramCountType EqualHeightHistogram<T>::_bin_distinct_count(const BinID index) const {
+  DebugAssert(index < _distinct_counts.size(), "Index is not a valid bin.");
+  return _distinct_counts[index];
+}
+
+template <typename T>
+HistogramCountType EqualHeightHistogram<T>::total_count() const {
+  return _total_count;
+}
+
+template <typename T>
+HistogramCountType EqualHeightHistogram<T>::total_distinct_count() const {
+  return std::accumulate(_distinct_counts.cbegin(), _distinct_counts.cend(), HistogramCountType{0});
```

Comment:
I'd expect this will get called more than once. Cache the total distinct count?

Reply:
Waiting for answers on same discussion above.